aws/update_other_tag.py

```diff
@@ -0,0 +1,208 @@
+    json_data.raise_for_status()
+    return json_data.json()
+
+
+def _resource_lookup(res_id):
+    """
+    Describe resource details of the given EC2 instance
+
+    TODO: support non-reserved instances; once we hit our monthly reservation,
+        the schema returned by ec2_client may change
+    TODO: support other types of resources (so far Cost Explorer has
+        only reported EC2 instances missing tags)
+    """
+
+    results = []
+    data = ec2_client.describe_instances(InstanceIds=[res_id, ])
+
+    if data:
+        logging.debug(data)
+        for reservation in data["Reservations"]:
+            for instance in reservation["Instances"]:
+                results.append(instance)
+
+    if len(results) == 0:
+        raise ValueError(f"Resource not found: {res_id}")
+
+    if len(results) > 1:
+        raise ValueError(f"Found multiple resources: {data}")
+
+    return results[0]
+
+
+def _get_cfn_stack_arns(stack_id):
+    """
+    Find all resources tagged with the same cloudformation stack
+    """
+
+    found = []
+    cfn_filter = {
+        "Key": "aws:cloudformation:stack-id",
+        "Values": [stack_id, ]
+    }
+
+    pager = tag_client.get_paginator("get_resources")
+    for page in pager.paginate(TagFilters=[cfn_filter, ]):
+        _list = page['ResourceTagMappingList']
+        for item in _list:
+            found.append(item["ResourceARN"])
+
+    return found
+
+
+def _get_sc_product_arns(product_id):
+    """
+    Find all resources tagged with the same provisioned product
+    """
+
+    found = []
+    sc_filter = {
+        "Key": "aws:servicecatalog:provisionedProductArn",
+        "Values": [product_id, ]
+    }
+
+    pager = tag_client.get_paginator("get_resources")
+    for page in pager.paginate(TagFilters=[sc_filter, ]):
+        _list = page['ResourceTagMappingList']
+        for item in _list:
+            if item["ResourceARN"] == product_id:
+                # This would cause an Invalid ARN error when applying the tag
+                continue
+            found.append(item["ResourceARN"])
+
+    return found
+
+
+def add_other_tag(res_id, other_value):
+    """
+    Update the CostCenterOther tag value on the given resource.
+
+    If the given resource is tagged with a cloudformation stack or a
+    provisioned service catalog product, then also apply the new tag
+    value to all other resources in the stack or product.
+
+    Also validate that the CostCenter tag is set to Other, and that the
+    new value for CostCenterOther is valid.
+    """
+
+    # look up resource id
+    item = _resource_lookup(res_id)
+    tags = _aws_user_tags(item['Tags'])
+    logging.info(f"Tags: {tags}")
+
+    # check that CostCenter is Other
+    if "CostCenter" not in tags:
+        raise ValueError(f"Resource {res_id} does not have a CostCenter tag")
+
+    cost_center = tags["CostCenter"]
+    if cost_center != "Other / 000001":
+        raise ValueError(f"Invalid CostCenter: {cost_center}")
```

Comment:
can't the `CostCenter` tag actually contain a valid cost center?  Maybe what you meant is that you are expecting a value of `Other` here however it's something else?